### PR TITLE
[rllib] MAML Agent

### PR DESCRIPTION
**Pull request by michaelzhiluo**
- Adds MAMLTrainer, meta RL algorithm

Tasks
- [x] Move Optimizer Code to Distributed Execution 
- [x] Benchmarks 
  - [x] `HalfCheetah-Rand-Direc`
  - [x] `Ant-Rand-Goal`
  - [x] `Pendulum-Mass`

For future PRs:
- [ ] Use Transform iter function for Inner Adaptation in Distributed Execution Plan
- [ ] Generalize meta-update graph for all policy architectures
- [ ] Minibatching for meta-update step
- [ ] Pytorch Implementation?

## Why are these changes needed?

Adds a new agent to RLLib

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/evaluation/rollout_worker.py [hunk 1]
```diff
@@ -503,7 +503,7 @@ def make_env(vector_index):
                 self.async_env, self.env, self.policy_map))
 
     @DeveloperAPI
-    def sample(self):
+    def sample(self, dataset_id=None):
```
Comment (ericl) [inline rllib/evaluation/rollout_worker.py:506]: Please don't change core APIs for adding new algorithms, instead use foreach_env / foreach_worker: https://docs.ray.io/en/master/rllib-training.html#curriculum-learning

File: rllib/agents/maml/maml.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    "batch_mode": "complete_episodes",
+    # Which observation filter to apply to the observation
+    "observation_filter": "NoFilter",
+    # Number of Inner adaptation steps for the MAML algorithm
+    "inner_adaptation_steps": 1,
+    # Number of MAML steps per meta-update iteration (PPO steps)
+    "maml_optimizer_steps": 5,
+    # Inner Adaptation Step size
+    "inner_lr": 0.1,
+<<<<<<< HEAD
```
Comment (ericl) [inline rllib/agents/maml/maml.py:58]: Can you fix these merge markers?

**Comment by ericl** [comment]
At a high level I like the design. @sven1977 do you want to review in detail?

File: rllib/agents/maml/maml.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    # If true, use the Generalized Advantage Estimator (GAE)
+    # with a value function, see https://arxiv.org/pdf/1506.02438.pdf.
+    "use_gae": True,
+    # GAE(lambda) parameter
+    "lambda": 1.0,
+    # Initial coefficient for KL divergence
+    "kl_coeff": 0.0005,
+    # Size of batches collected from each worker
+    "rollout_fragment_length": 200,
+    # Stepsize of SGD
```
Comment (sven1977) [inline rllib/agents/maml/maml.py:31]: This looks very large, no?

File: rllib/agents/maml/maml.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    # @mluo: TODO
+    assert config["framework"] != "torch"
+    return MAMLTFPolicy
+
+
+def validate_config(config):
+    if config["inner_adaptation_steps"] <= 0:
+        raise ValueError("Inner Adaptation Steps must be >=1.")
+    if config["maml_optimizer_steps"] <= 0:
+        raise ValueError("")
```
Comment (sven1977) [inline rllib/agents/maml/maml.py:210]: Can you fill this in?

File: rllib/agents/maml/maml.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+
+def get_policy_class(config):
+    # @mluo: TODO
+    assert config["framework"] != "torch"
+    return MAMLTFPolicy
+
+
+def validate_config(config):
+    if config["inner_adaptation_steps"] <= 0:
```
Comment (sven1977) [inline rllib/agents/maml/maml.py:207]: Can you add that pytorch is not supported?
```
if config["framework"] == "torch": ...
```

File: rllib/agents/maml/maml_tf_policy.py [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+                self.kl_coeff_val[i] *= 0.5
+            elif kl > 1.5 * self.kl_target:
+                self.kl_coeff_val[i] *= 2.0
+        print(self.kl_coeff_val)
+        self.kl_coeff.load(self.kl_coeff_val, session=self.get_session())
+        return self.kl_coeff_val
+
+
+def maml_optimizer_fn(policy, config):
+    if not config["worker_index"]:
```
Comment (sven1977) [inline rllib/agents/maml/maml_tf_policy.py:402]: Can you add a comment here to explain, why we are using two different optimizer types?

File: rllib/tuned_examples/maml/ant-rand-goal-maml.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        kl_target: 0.01
+        kl_coeff: 0.0005
+        num_workers: 32
+        num_gpus: 1
+        inner_lr: 0.03
+        explore: True
+        clip_actions: False
+        model:
+            fcnet_hiddens: [64, 64]
+            free_log_std: True
```
Comment (sven1977) [inline rllib/tuned_examples/maml/ant-rand-goal-maml.yaml:26]: Can you add a newline here at the EOF?

File: rllib/tuned_examples/maml/ant-rand-goal-maml.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
+ant-rand-goal-maml:
```
Comment (sven1977) [inline rllib/tuned_examples/maml/ant-rand-goal-maml.yaml:1]: Nice!

File: rllib/agents/maml/tests/test_maml.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        ray.init()
+
+    @classmethod
+    def tearDownClass(cls):
+        ray.shutdown()
+
+    def test_maml_compilation(self):
+        """Test whether a MAMLTrainer can be built with all frameworks."""
+        config = maml.DEFAULT_CONFIG.copy()
+        config["num_workers"] = 1  # Run locally.
```
Comment (sven1977) [inline rllib/agents/maml/tests/test_maml.py:23]: Remove the "Run locally" comment.

File: rllib/tuned_examples/maml/pendulum-mass-maml.yaml [hunk 10]
```diff
@@ -0,0 +1,26 @@
+pendulum-mass-maml:
+    env: ray.rllib.examples.env.pendulum_mass.PendulumMassEnv
+    run: MAML
+    stop:
```
Comment (sven1977) [inline rllib/tuned_examples/maml/pendulum-mass-maml.yaml:4]: If these are actually learning, could you add some decent stopping criteria here? Like `episode_reward_mean` and `training_iteration` (or better `timesteps_total`).

File: rllib/tuned_examples/maml/halfcheetah-rand-direc-maml.yaml [hunk 11]
```diff
@@ -0,0 +1,25 @@
+halfcheetah-rand-direc-maml:
+    env: ray.rllib.examples.env.halfcheetah_rand_direc.HalfCheetahRandDirecEnv
+    run: MAML
+    stop:
+        training_iteration: 2001
```
Comment (sven1977) [inline rllib/tuned_examples/maml/halfcheetah-rand-direc-maml.yaml:5]: If these are actually learning, could you add some decent stopping criteria here? Like episode_reward_mean and training_iteration (or better timesteps_total).

File: rllib/tuned_examples/maml/ant-rand-goal-maml.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    env: ray.rllib.examples.env.ant_rand_goal.AntRandGoalEnv
+    run: MAML
+    stop:
+        training_iteration: 1001
```
Comment (sven1977) [inline rllib/tuned_examples/maml/ant-rand-goal-maml.yaml:5]: If these are actually learning, could you add some decent stopping criteria here? Like episode_reward_mean and training_iteration (or better timesteps_total).

File: rllib/examples/env/ant_rand_goal.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
+import numpy as np
```
Comment (sven1977) [inline rllib/examples/env/ant_rand_goal.py:1]: Really cool!
Could you add some comments/docstrings to the methods of these Envs? Or, if these are overrides from parent, could you annotate them using our `@override([]parent)` decorator?
These seem very useful as templates for users to create their custom Env-distributions.

**Review by sven1977 — CHANGES_REQUESTED** [review]

**Comment by sven1977** [comment]
Looks very good! Thanks @michaelzhiluo . Just some smaller items remaining (comments and stuff).

File: rllib/agents/maml/maml.py (hunk 3 above)
Comment (michaelzhiluo) [reply rllib/agents/maml/maml.py:31]: True, Pendulum-Mass and Antrandgoal both have an episode horizon of length 200, hence it defaults to 200. (each worker env is stepping through 1 episode)

File: rllib/examples/env/ant_rand_goal.py (hunk 13 above)
Comment (michaelzhiluo) [reply rllib/examples/env/ant_rand_goal.py:1]: Ah they do not override, the Parents dont have sample_tasks, get_task, set_task or step method. However,

File: rllib/tuned_examples/maml/ant-rand-goal-maml.yaml (hunk 8 above)
Comment (michaelzhiluo) [reply rllib/tuned_examples/maml/ant-rand-goal-maml.yaml:1]: :)

File: rllib/tuned_examples/maml/ant-rand-goal-maml.yaml (hunk 12 above)
Comment (michaelzhiluo) [reply rllib/tuned_examples/maml/ant-rand-goal-maml.yaml:5]: The episode_reward_mean is the pre adaptation return. The post adaptation return is episode_reward_mean_adapt_[#]. For timesteps_total, I think training_iteration is a better criteria b/c that is how many meta-update steps it took to reach a good meta-policy. (one metaupdate per iteration)

File: rllib/tuned_examples/maml/halfcheetah-rand-direc-maml.yaml (hunk 11 above)
Comment (michaelzhiluo) [reply rllib/tuned_examples/maml/halfcheetah-rand-direc-maml.yaml:5]: Above comment^^

File: rllib/tuned_examples/maml/pendulum-mass-maml.yaml (hunk 10 above)
Comment (michaelzhiluo) [reply rllib/tuned_examples/maml/pendulum-mass-maml.yaml:4]: Above comment^^

**Comment by michaelzhiluo** [comment]
> Looks very good! Thanks @michaelzhiluo . Just some smaller items remaining (comments and stuff).

Thank you for going through this very long PR! I agree with all of your proposed changes, except for the `training-iterations` for the yaml files.

File: rllib/agents/maml/tests/test_maml.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+        num_iterations = 1
+
+        # Test for tf framework (torch not implemented yet).
+        for _ in framework_iterator(config, frameworks=("tf")):
+            trainer = maml.MAMLTrainer(
+                config=config,
+                env="ray.rllib.examples.env.pendulum_mass.PendulumMassEnv")
+            for i in range(num_iterations):
+                trainer.train()
+            check_compute_action(trainer, include_prev_action_reward=True)
```
Comment (sven1977) [inline rllib/agents/maml/tests/test_maml.py:35]:
```suggestion
            check_compute_single_action(trainer, include_prev_action_reward=True)
```

File: rllib/agents/maml/tests/test_maml.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
+import unittest
+
+import ray
+import ray.rllib.agents.maml as maml
+from ray.rllib.utils.framework import try_import_tf
+from ray.rllib.utils.test_utils import framework_iterator, check_compute_action
```
Comment (sven1977) [inline rllib/agents/maml/tests/test_maml.py:6]:
```suggestion
from ray.rllib.utils.test_utils import framework_iterator, check_compute_single_action
```

File: rllib/agents/maml/tests/test_maml.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+        num_iterations = 1
+
+        # Test for tf framework (torch not implemented yet).
+        for _ in framework_iterator(config, frameworks=("tf")):
+            trainer = maml.MAMLTrainer(
+                config=config,
+                env="ray.rllib.examples.env.pendulum_mass.PendulumMassEnv")
+            for i in range(num_iterations):
+                trainer.train()
+            check_compute_single_action(trainer, include_prev_action_reward=True)
```
Comment (sven1977) [inline rllib/agents/maml/tests/test_maml.py:36]:
```suggestion
            check_compute_single_action(
                trainer, include_prev_action_reward=True)
```

**Review by sven1977 — APPROVED** [review]
Very nice! Thanks @michaelzhiluo !

**Pull request merged**